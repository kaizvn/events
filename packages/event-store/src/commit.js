/* @flow */
import { createError } from 'micro-boom';
import { json } from 'micro';
import redis from 'redis';

import runLua from './runLua';

type Event = {
  type: string,
  payload: mixed,
};

const client = redis.createClient();

const commit = async (redisClient, event: Event) => {
  const lua = `
    local event = ARGV[1];
    local counter = redis.call('HLEN', KEYS[1]) + 1;

    redis.call('HSET', KEYS[1], counter, event);
    redis.call('PUBLISH', 'events', counter .. ':' .. event);
    return counter;
  `;

  return runLua(redisClient, lua, {
    keys: ['events'],
    argv: [JSON.stringify(event)],
  });
};

type TransformFunction = (any, http$IncomingMessage) => Event;

export default (transformReq: TransformFunction) => async (req: any) => {
  const body = await json(req);
  const event = transformReq(body, req);

  if (typeof event.type !== 'string') {
    throw createError(422, 'Missing type');
  }

  const id = await commit(client, event);

  return { ...event, id };
};
