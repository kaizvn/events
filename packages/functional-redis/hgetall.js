const promisify = require('es6-promisify');
const curry = require('curry');

module.exports = promisify((key, client, cb) => client.hgetall(key, cb));
