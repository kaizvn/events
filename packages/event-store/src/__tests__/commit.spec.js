/* @flow */
import flushdb from 'redis-functional/flushdb';
import hgetall from 'redis-functional/hgetall';

import commit from '../commit';
import redisClient from '../redis-client';

describe('commit endpoint', () => {
  beforeEach((done) => {
    flushdb(redisClient, done);
  });

  it('should commit', async () => {
    const transformReq = jest.fn(obj => ({
      type: 'transformed_type',
      payload: { key: 'transformed_value' },
      meta: {
        client: 'test_client',
      },
    }));

    const req = {
      body: {
        type: 'test',
        payload: { key: 'value' },
      },
    };

    const service = commit(transformReq);

    const actual = await service(req);
    expect(actual).toEqual({
      id: 1,
      type: 'transformed_type',
      payload: { key: 'transformed_value' },
      meta: {
        client: 'test_client',
      },
    });

    expect(transformReq).toBeCalledWith(req.body, req);
    expect(await hgetall('events', redisClient)).toEqual({
      '1': JSON.stringify({
        type: 'transformed_type',
        payload: { key: 'transformed_value' },
        meta: {
          client: 'test_client',
        },
      })
    })
  });
});
