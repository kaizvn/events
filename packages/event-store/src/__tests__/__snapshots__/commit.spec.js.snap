// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commit endpoint should throw 422 if no type is sent 1`] = `[Error: Missing type]`;
